### PR TITLE
Web Inspector: Create All Fonts Tab in Font Section

### PR DESCRIPTION
**Pull request by francescorn**
#### b6f332bf9b5fc5c853b02114cc98cdf1bd489272
<pre>
Web Inspector: Create All Fonts Tab in Font Section
<a href="https://bugs.webkit.org/show_bug.cgi?id=260127">https://bugs.webkit.org/show_bug.cgi?id=260127</a>

Reviewed by NOBODY (OOPS!).

Co-authored by Brandon Stewart.

We created a Custom Fonts tab to display all of the Custom Fonts in the DOM, and to auto refresh to account for adjusting webpages such as slideshows.
There are parameters font-family name, stretch, style, and loaded, and can sort by each parameter as well.
If there are no custom fonts in the DOM, &quot;No Custom Fonts&quot; will be displayed.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getCustomFonts):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.get customFonts):
(WI.DOMNodeStyles.prototype.refresh):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype._sortDataGrid.comparator):
(WI.FontDetailsPanel.prototype._sortDataGrid):
</pre>
https://github.com/WebKit/WebKit/commit/b6f332bf9b5fc5c853b02114cc98cdf1bd489272

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54582 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34001 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5316 "Hash b6f332bf for PR 29525 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56884 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5318 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/59/builds/5316 "Hash b6f332bf for PR 29525 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56677 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3456 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/47947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54088 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/5318 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30965 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66386 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12656 "Passed tests") |

### CODE REVIEW

**Comment by francescorn** [comment]
This is what it looks like now:
Screenshot%202024-06-04%20at%205 53 39%E2%80%AFPM
. Link to YouTube video: https://youtu.be/LSmwW0M3BG8.

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 1]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
     return fontFamilyNames;
 }
 
+Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::CSS::FontFace>>> InspectorCSSAgent::getCustomFonts()
+{
+    auto customFonts = JSON::ArrayOf<Inspector::Protocol::CSS::FontFace>::create();
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
+
+    for (auto* document : domAgent->documents()) {
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: there's no guarantee that the `domAgent` will exist here, so id either wrap this entire `for` in a `if (auto* domAgent = m_instrumentingAgents.persistentDOMAgent()) { ... }` or add a `return` here in a `if (!domAgent)`

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 2]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+        auto iterator = document->fonts().get().createIterator(document->scriptExecutionContext());
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: you can do `document->fonts()->` instead of calling `.get()`

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 3]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+
+        while (auto font = iterator.next()) {
+
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: Style: unnecessary newline

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 4]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+            String status = ""_s;
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: no need to assign to `""_s` as `status` will always be set below

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 5]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+                break;
+            case FontFace::LoadStatus::Error:
+                status = "error"_s;
+                break;
+            case FontFace::LoadStatus::Loaded:
+                status = "loaded"_s;
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: this is unnecessary since you've already covered all the possible values of `FontFace::LoadStatus`

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 6]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
+
+    for (auto* document : domAgent->documents()) {
+        auto iterator = document->fonts().get().createIterator(document->scriptExecutionContext());
+
+        while (auto font = iterator.next()) {
+
+            String status = ""_s;
+
+            switch (font.get()->status()) {
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: you can do `font->` instead of calling `.get()`

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 7]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+            .setFamily(font.get()->family())
+            .setWeight(font.get()->weight())
+            .setStretch(font.get()->stretch())
+            .setUnicodeRange(font.get()->unicodeRange())
+            .setFeatureSettings(font.get()->featureSettings())
+            .setDisplay(font.get()->display())
+            .setSizeAdjust(font.get()->sizeAdjust())
+            .setLoaded(status)
+            .release();
+            customFonts->addItem(result);
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: this should be `customFonts->addItem(WTFMove(result))` since it's no longer needed (i.e. we can relinquish ownership)

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 8]
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
...
+            case FontFace::LoadStatus::Loaded:
+                status = "loaded"_s;
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
+            }
+
+            Ref<Inspector::Protocol::CSS::FontFace> result = Inspector::Protocol::CSS::FontFace::create()
+            .setFamily(font.get()->family())
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: you can do `font->` instead of calling `.get()`

Style: also please indent all of these
```cpp
auto fontFace = Inspector::Protocol::CSS::FontFace::create()
    .setFamily(font->family())
    .setWeight(font->weight())
    ...
    .release();
customFonts->addItem(WTFMove(fontFace));
```

File: Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js [hunk 9]
```diff
@@ -156,6 +156,10 @@ localizedStrings["All Changes"] = "All Changes";
 localizedStrings["All Events @ Event Breakpoint"] = "All Events";
 /* Break (pause) on all exceptions */
 localizedStrings["All Exceptions @ JavaScript Breakpoint"] = "All Exceptions";
+/* Font Details Panel */
+localizedStrings["No Custom Fonts @ Font Details Sidebar Property"] = "No Custom Fonts";
```
Comment (dcrousso) [inline Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js]: please do not manually edit this file.  it's autogenerated by running
```
./Tools/Scripts/extract-localizable-js-strings --utf8 ./Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js ./Source/WebInspectorUI/UserInterface
```

File: Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js [hunk 10]
```diff
@@ -135,6 +137,8 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
     get usedCSSVariables() { return this._usedCSSVariables; }
     get allCSSVariables() { return this._allCSSVariables; }
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js]: Style: unnecessary newline

File: Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js [hunk 11]
```diff
@@ -340,6 +344,15 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
         target.CSSAgent.getInlineStylesForNode.invoke({nodeId: this._node.id}, wrap.call(this, fetchedInlineStyles, fetchedInlineStylesPromise));
         target.CSSAgent.getComputedStyleForNode.invoke({nodeId: this._node.id}, wrap.call(this, fetchedComputedStyle, fetchedComputedStylesPromise));
 
+
+        if (InspectorBackend.hasCommand("CSS.getCustomFonts")) {
+            target.CSSAgent.getCustomFonts((error, fontFamilyNames) =>{
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:347]: calling this every time we create/update a `WI.DOMNodeStyles` seems like way too much since it gets all fonts for the entire page, not just for a specific node (which is really what `WI.DOMNodeStyles` is about

id suggest moving this to `WI.CSSManager` and having an event from the backend (e.g. `CSS.fontsChanged`) that then triggers this to be called again, as that way we dont call it more than once for the same data

Style: missing space `=> {`

File: Source/JavaScriptCore/inspector/protocol/CSS.json [hunk 12]
```diff
@@ -462,6 +477,14 @@
                 { "name": "fontFamilyNames", "type": "array", "items": { "type": "string" }, "description": "Supported system font families." }
             ]
         },
+        {
+            "name": "getCustomFonts",
+            "description": "Returns all loaded fonts (system and custom).",
```
Comment (dcrousso) [inline Source/JavaScriptCore/inspector/protocol/CSS.json:482]: if this also includes system fonts, why is it called `getCustomFonts`?  wouldn't something like `getActiveFonts` or `getKnownFonts` be better?

also if this includes system fonts do we still need `getSupportedSystemFontFamilyNames`?

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 13]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
             this._fontVariationsGroup.rows = [emptyRow];
         }
+
+        // Custom Font Properties
+
+        // Display message when we have no custom fonts.
+        if (!this.nodeStyles.customFonts.length) {
+            this._customFontsRow.showEmptyMessage();
+            this._customFontsRow.dataGrid = null;
+            return;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: we typically dont add early `return` in functions like this as it makes it harder for new things to be added at the end in the future.  as such, id put everything after this in an `} else { ... }` (or more ideally flip it around and have the positive check be first with the error case in the `else`)

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 14]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+                aligned: "centered"
+            },
+            loaded: {
+                title : WI.UIString("Loaded"),
+                sortable: true,
+                aligned: "centered"
+            }
+        };
+
+        this._dataGrid = new WI.DataGrid(fontColumns);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: this needs a more unique name like `_allFontsDataGrid`

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 15]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+        this._dataGrid.inline = true;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: im not sure this is correct

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 16]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+        this._dataGrid.sortColumnIdentifier = "family";
+        this._dataGrid.sortOrder = WI.DataGrid.SortOrder.Ascending;
+        this._dataGrid.createSettings("font-details-panel-all-fonts");
+        this._dataGrid.addEventListener(WI.DataGrid.Event.SortChanged, this._sortDataGrid, this);
+    }
+
+
+    _sortDataGrid()
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: ditto with a more unique name

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 17]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+    {
+        var sortColumnIdentifier = this._dataGrid.sortColumnIdentifier;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: Style: please use `let`

ditto below

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 18]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
 
+    _createDataGrid()
+    {
+        // All Fonts Properties
+        let fontColumns = {
+            family: {
+                title : WI.UIString("Family"),
+                sortable: true,
+                aligned: "centered",
+                width: "50%"
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: Style: missing trailing `,`

ditto below

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 19]
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+
+        this.element.appendChild(customFontsSection.element);
     }
 
     // Private
 
+    _createDataGrid()
+    {
+        // All Fonts Properties
+        let fontColumns = {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: Style: this can be `const` since it doesn't change between Web Inspector sessions

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 20]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+            this._customFontsRow.showEmptyMessage();
+            this._customFontsRow.dataGrid = null;
+            return;
+        }
+
+        // We can have no dataGrid if we navigate to a page
+        // with no custom fonts. This will trigger it to force
+        // rebuild the table.
+        if (!this._customFontsRow.dataGrid)
+            this._customFontsRow.dataGrid = this._dataGrid;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: rather than keeping the `_allFontsDataGrid` as a member variable, i think instead we should `return` it from `_createAllFontsDataGrid` and then have that be called here
```js
this._customFontsRow.dataGrid ||= this._createAllFontsDataGrid();
this._customFontsRow.dataGrid.removeChildren();
...
```

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 21]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+
+        this._customFontsRow.dataGrid.removeChildren();
+
+        for (let customFont of this.nodeStyles.customFonts) {
+
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:143]: Style: unnecessary whitespace

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 22]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+            // The backend can return empty font families, which we should ignore.
+            if (customFont.family === "")
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: should we be filtering this in the backend?  im not sure i see a benefit in sending a font to the frontend if there's nothing to display for the name

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 23]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
...
+                continue;
+
+            let dataGridNode = new WI.DataGridNode({family: customFont.family, weight: customFont.weight, stretch: customFont.stretch, style: customFont.stretch, loaded: customFont.loaded});
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js]: Style: please separate each property onto a separate line for better readability

File: Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js [hunk 24]
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
 
             this._fontVariationsGroup.rows = [emptyRow];
         }
+
+        // Custom Font Properties
+
+        // Display message when we have no custom fonts.
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:127]: Style: these kinds of comments are not really useful as it's pretty straightforward from reading the code what this will do

File: Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js [hunk 25]
```diff
@@ -340,6 +344,15 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
...
+                if (error)
+                    return;
+                this._customFonts = fontFamilyNames;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:350]: i think this should dispatch some sort of `WI.CSSManager.Event.FontsChanged` so that listeners (e.g. `WI.FontDetailsPanel`) can listen to it and update accordingly

File: Source/JavaScriptCore/inspector/protocol/CSS.json [hunk 26]
```diff
@@ -268,6 +268,21 @@
...
+          "description": "A FontFace",
+          "properties": [
+              { "name": "family", "type": "string", "description": "Font Family Name"},
+              { "name": "weight", "type": "string", "description": "Font Stretch"},
+              { "name": "stretch", "type": "string", "description": "Font Family Name"},
+              { "name": "unicodeRange", "type": "string", "description": "Font Unicode Range"},
+              { "name": "featureSettings", "type": "string", "description": "Font Feature Settings"},
+              { "name": "display", "type": "string", "description": "Font Display"},
+              { "name": "sizeAdjust", "type": "string", "description": "Font Size Adjust"},
+              { "name": "loaded", "type": "string", "description": "Font Loaded"}
```
Comment (dcrousso) [inline Source/JavaScriptCore/inspector/protocol/CSS.json:283]: is this actually a useful piece of information for the developer?

are there any actions they can take given this info?  or is it purely informative?

if the latter, id suggest not including it as Web Inspector usually tries to only provide information that is actionable, as otherwise it's potentially confusing

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 1 above)
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: im not entirely sure what benefit there is in grabbing fonts for _all_ documents, as i dont think fonts are usable in situations like across `<iframe>`

perhaps we should adjust this to be something that gets the lists of fonts given a node (or more ideally a document node)?  that way we can show only the list of usable fonts for the selected node in the Fonts details sidebar panel

**Review by dcrousso — CHANGES_REQUESTED** [review]
overall the code itself looks mostly fine with a few bugs and stylistic things to fix

conceptuatlly, im a bit concerned with the idea of this feature, as i think all of the other information in the Font details sidebar panel is related to the selected node, not general to the entire page (i.e. it could confuse the developer into incorrectly thinking that all the listed fonts are being used by the selected node)

also, given the fact that `document.fonts` exists in JS, im not sure what this provides to the developer beyond them evaluating that in the Console themselves (i.e. Web Inspector aims to provide unique functionality/information/etc. that cannot be accessed elsewhere)

id much rather see us be able to split the list of all fonts into ones used by the selected node (and ideally with the specific unicode ranges being used from the font) and those that aren't, as that way the developer can both understand what they are using as well as what it's possible to use (which could be fed into `font`/`font-family` autocomplete as well)

**Comment by yisibl** [comment]
@francescorn Can we display the `@font-face` definition of a custom font?

Like this:
image

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 1 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I replaced with this:
```
if (auto* domAgent = m_instrumentingAgents.persistentDOMAgent()) {
        for (auto* document : domAgent->documents()) {
```

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 2 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I removed the `.get`s

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 3 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I removed the new line

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 4 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I removed status here.

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 5 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I removed the assert.

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 6 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I removed .get from everything I added it.

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 7 above)
Comment (francescorn) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: I replaced ```customFonts->addItem(result);``` with ```customFonts->addItem(WTFMove(result));```

File: Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js (hunk 9 above)
Comment (francescorn) [reply Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js]: I ran the command and got:
```/* Message shown when there are no custom fonts. */
localizedStrings["No Custom Fonts @ Font Details Sidebar Property"] = "No Custom Fonts";```
populated.
